File: competitive_programming/leetcode/python/329_longest_increasing_path_in_a_matrix.py
from typing import List
class Solution:
    def __init__(self):
        self.n = 0
        self.m = 0
        self.maximum = 0
        self.direction = [0, -1, 0, 1, 0]

    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        self.m = len(matrix)
        self.n = len(matrix[0])
        memo = [[0 for _ in range(self.n)] for _ in range(self.m)]
        for i in range(self.m):
            for j in range(self.n):
                self.maximum = max(self.maximum, self.dfs(i, j, matrix, memo))
        return self.maximum           

    def dfs(self, ri: int, rj: int, matrix: List[List[int]], memo: List[List[int]]) -> int:
        if memo[ri][rj] > 0:
            return memo[ri][rj]
        memo[ri][rj] += 1
        for k in range(4):
            i = ri + self.direction[k]
            j = rj + self.direction[k + 1]
            if i < 0 or i >= self.m or j < 0 or j >= self.n or matrix[i][j] <= matrix[ri][rj]:
                continue
            memo[ri][rj] = max(memo[ri][rj], self.dfs(i, j, matrix, memo) + 1)
        return memo[ri][rj]

solution = Solution()
matrix = [
    [9, 9, 4],
    [6, 6, 8],
    [2, 1, 1]
]
print(solution.longestIncreasingPath(matrix))

File: competitive_programming/leetcode/cpp/329_longest_increasing_path_in_a_matrix.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        m =  matrix.size();
        n =  matrix[0].size();
        int maximum = 0;
        vector<vector<int>> memo(m, vector<int>(n));
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                maximum = max(maximum, dfs(i, j, memo, matrix)); 
            }
        }
        return maximum;
    }

private:
    int m, n;
    int direction [5] = {0, -1, 0, 1, 0}; 

    int dfs(int ri, int rj, vector<vector<int>>& memo, const vector<vector<int>>& matrix) {
        if (memo[ri][rj] > 0) {
            return memo[ri][rj];
        }
        memo[ri][rj] = 1;
        for (int k = 0; k < 4; ++k) {
            int i = ri + direction[k];
            int j = rj + direction[k + 1];
            if (i < 0 || i >= m || j < 0 || j >= n || matrix[i][j] <= matrix[ri][rj]) {
                continue;
            }
            memo[ri][rj] = max(memo[ri][rj], dfs(i, j, memo, matrix) + 1);
        }
        return memo[ri][rj]; 
    }
};

int main() {
    Solution *solution = new Solution();
    vector<vector<int>> matrix;

    matrix = {
        {9, 9, 4},
        {6, 6, 8},
        {2, 1, 1}
    };
    cout << solution->longestIncreasingPath(matrix) << "\n";

    matrix = {
        {3, 4, 5},
        {3, 2, 6},
        {2, 2, 1}
    };
    cout << solution->longestIncreasingPath(matrix) << "\n";

    matrix = {{1}};
    cout << solution->longestIncreasingPath(matrix) << "\n";
}